### DISCUSSION TITLE
Proposal: Floor Plan Entity

### DISCUSSION THREAD

**Discussion by Lash-L** [home-assistant/architecture#950] [Entity Models]
per comments on https://github.com/home-assistant/core/pull/93633 and https://github.com/home-assistant/architecture/discussions/913#discussioncomment-6043647

## Floor Plans
Floor plans would be specified by an integration with a list of rooms

A room would contain the following information:
- Name
- Bounds ( a list of tuple of floats that would represent the bounds of each room.)
- Optional background image
- clickable action (?) (Optional) - I think it would be nice if I could click on a room and see all actions I can do on that room, i.e. if I click on my living room, I could see the option to vacuum it.  Alternatively, if my solution for room to area mapping is accepted, clicking on a room would open it's area's page. If it doesn't have an area set, it would open a dialog to do that or to create a new area

This would probably not be updated, or not be updated frequently as it will likely be a time consuming task. 

### Floor Plan Object
There would be a secondary object called a floor plan object. This object can be updated if desired. In the example of a Vacuum cleaner floor plan, this is where you would put the vacuum cleaner.

This would contain the following information:
- x,y location (in the same coordinate space as the room bounds)
- if it can be updated a function to update the location
- Image/icon
- name
- transparency
- size
- Other properties kwargs that may be needed for updating.
- Link to the Entity - Click on it and it opens the entity page or just performs an action (i.e. switch)
These Floor plan objects could be added through a service (or potentially a ui option). This way, you could use the floor plan provided by your vacuum, and add objects in your house throughout the floor plan. i.e. if I wanted to add a button for lights, I could do that and I would give the location I want it to go on the floor plan. Then when I click on it, it would turn my lights on or off.

They should also be able to be disabled/enabled.

## Unknowns:

- It could be nice to sync the rooms with the Area entities(this would have to be done by the user) but the issue is there is a potentially many to many mapping and 2 rooms may represent one area or two areas may represent one room.
Edit: My solution to that problem is that it should probably be a many to one relationship. Many rooms can be part of one area, but not the other way around, if it is the other way around, more areas should be created or rooms should be deleted.

**Comment by allenporter** [comment]
Nice exploration of this topic.  I was remembering that since we started discussing this, there was also a new image entity that was introduced. I wonder if the image entity could be a basis for this? (either as an initial starting place or as an end solution with additional attributes associated with images down the line).  Could the integration render the image with the objects on it itself? 

That may not work, but my line of thinking here was to consider separating out these concerns can help us get to as simple as possible proposal:
- rendering the floor plan and any objects within the floor plan
- information needed for interactive components (e.g. information needed by the frontend to allow control of a device related to the image shown)
- if needed, representation of metadata needed for other control e.g. automation, state,  "the vacuum is in the living room, so turn off the living room motion detector" if there is an actual use case for something like this

**Reply by Lash-L** [reply]
I think having the floor map itself contained within an Image Entity is no problem, it is just the interactions that get tricky that I don't know about. Being able to do multiple actions depending on what pixel you click on may be far outside of the scope of the Image entity, I'll have to do some exploration into the codebase for Image

**Comment by gjohansson-ST** [comment]
I have to say reading this it doesn't sound like an entity.

**Reply by Lash-L** [reply]
> I have to say reading this it doesn't sound like an entity.

Why not? I would say it's very similar to an image entity for example - and the information would have to be provided by an integration (or a user manually inputting)

What would you say it should be if not an entity?

**Reply by EnochPrime** [reply]
I was thinking the same, that it doesn't really sound like an entity. This is because it is sort of a hybrid of the functionality of an area and a zone which are not entities. I can understand though that integrations provide data through entities so it could make sense to use that existing structure.

**Reply by allenporter** [reply]
@Lash-L this may seem super obvious or not needed, but maybe it is worth spelling out the implicit use cases to make them explicit. What are the end user problems we're solving with floor plans synced to Home Assistant?

When i look at functionality available in vacuum apps i see:
- showing progress of a vacuum on a floor plan (for example, the path taken on the floor plan which shows the area covered and not yet covered)
- ability to multi-select rooms to control the clean targets
- zones, to direct cleaning to a specific sub-area within a room
- More generally when considering use cases for floor plans, I gather from the proposal it involves things like putting icons on the UI and interacting with the entity (similar to a ui card configuration).

It seems like (1) can be done with an image. (2) can be done with a muli-select drop down discussed in the other arch discussion (3) i don't ever use so not sure if its the most important thing or not (4) seems like it needs some kind of area coordinates associated with a floorplan image, but other requirements there could be left to the UI configuration (e.g. exactly where in an area an entity is may just be a display thing, and not necessarily needed to be known by entities)

I may be off here.  Are there more use cases or missing requirements here? My suggestion is start with case (1) with an image entity drawing progress of the vacuum on the floor plan. The other arch discussion talks about how to sync areas which is cool. The last missing bit is how to associate areas with coordinates on an image, which maybe could also just be a UI-only thing.

**Reply by Lash-L** [reply]
Sorry for such a long delay to respond. I'm redoing my bathroom and my laptop broke, so I haven't had a lot of time/ ability to do HA stuff.

I think all of those features are nice, but where this would really thrive in my opinion, is the ability to add things that don't relate to your vacuum. Getting a floor plan of your house that you can use as a dashboard for controlling your home - without having to take measurements and figure out how to draw it/add it to HA will be really helpful for a lot of people I think

But in terms of what this needs to do - I think a good starting/ needed starting is the ability to draw the image of the floor plan and keep up with it, which isn't too hard and we currently do it in the custom integration for Roborock. But it gets more tricky when adding in the ability to control the vacuum with the floor plan and I think that is where a new entity/ feature directly in HA core comes in.

**Comment by marksteward** [comment]
It's worth considering how it might interact with ha-floorplan. They have a note about how time-consuming it is, and vacuum data could help that.

They use SVG elements for rendering, so maybe it makes sense to translate the map data into SVG shapes/paths and attach that to areas or devices? This could all then be exposed in JSON over the websocket interface, like zones are.

**Comment by HarvsG** [comment]
I think there are a whole set of use cases for HA to be floorplan-aware (and where devices are located on the floor-plan) (whether or not an entity is the right implementation):
- Robot vacuums
- Bluetooth device tracking https://github.com/agittins/bermuda
- mmWave presence tracking https://github.com/EverythingSmartHome/everything-presence-addons
- Prescence tracking using other protocols like Zigbee and wifi https://www.philips-hue.com/en-us/support/article/motionawaretm--transform-your-hue-lights-into-motion-sensors/000011
- Optimal delay calculations in multi-room audio sync https://discord.com/channels/753947050995089438/1451524060537421834
- Speaker location and directionality for voice context
- Beautiful digital-twin dashboards
- VR/MR UIs https://www.reddit.com/r/homeassistant/comments/1cj7d7z/showcase_of_my_mixed_reality_dashboard_for_home/

**Reply by AJediIAm** [reply]
This proposal is implemented by https://www.home-assistant.io/docs/organizing/areas/

With the new https://www.home-assistant.io/blog/2025/12/03/release-202512/#purpose-specific-triggers-and-conditions, automations can trigger based on a device type (like occupancy) in a area of floor.

There are various options for a floorplan dashboard such as https://experiencelovelace.github.io/ha-floorplan/

**Reply by marksteward** [reply]
We're aware of areas, they're simply containers, and don't cover any of the geographical/location features discussed above.

**Reply by HarvsG** [reply]
Agreed, although I do think that the floorplan may be better suited as an extension of the areas object rather than a distinct entity
